rfd/XXXX-securing-amis.md

```diff
@@ -0,0 +1,225 @@
+partially obviate the need to move the legacy images from the `gravitational`
+AWS account. 
+
+As per the existing AMI builds, AMIs will be constructed as part of the tag build, and
+made public on promotion.
+
+#### **Post-build validation using Amazon Inspector**
+
+While the CIS-hardening tooling on Amazon requires royalties and subscriptions,
+the Amazon Inspector tool contains CIS benchmarks appears to be royalty free. 
+
+Amazon Inspector can't examine AMIs directly, but it is possible to 
+ 1. spin up an EC2 instance based on the given AMI, 
+ 2. install the AWS Inspector agent onto the Instance,
+ 3. run an Inspector assessment over the instance,
+ 4. interrogate the resulting list of findings to decide if a build passes
+    or not
+
+This idea is based on a [Continuous Vulnerability Assessment](https://aws.amazon.com/blogs/security/how-to-set-up-continuous-golden-ami-vulnerability-assessments-with-amazon-inspector/) example from the AWS security blog, minus the scheduled lambda to trigger the assessment.
+
+#### **Ongoing Scanning**
+
+We should mirror our approach for OCI container Images ([See RFD-0112](./0112-securing-docker-images.md)), which boils down to:
+
+ * Routine (e.g. daily) scanning of the most recent release of each 
+   supported Teleport version (i.e. 3 most recent major versions). 
+ * Detected issues written to the `teleport.e` GitHub Security Issues 
+   tab
+ * Any further processing to be handled by our Panther SIEM
+
+The `trivy` scanner is [known to support scanning AMIs](https://aquasecurity.github.io/trivy/v0.35/docs/vm/aws/), 
+and we already it for OCI container images and other resources. We should
+use it for our AMIs as well.
+
+#### **AMI Cleanup**
+
+There is a default 1114 limit of public AMIs allowed in an AWS region. While this 
+can be raised on request, it seems reasonable to periodically clean up our public
+images.
+
+Proposed deletion criteria:
+ * non-release builds (i.e. anything with a semver suffix) older than 6 months
+ * release builds with Teleport major version <= 5 (updated over time )
+
+### **Implementation Plan**
+
+  * **Phase 1:** 
+    * Hardened base image, 
+    * re-based AMIs, 
+    * publishing to `teleport-prod`
+  * **Phase 2:**
+    * Post-build validation
+  * **Phase 3:**
+    * Ongoing scanning
+  * **Phase 4:**
+    * AMI Cleanup
+
+I expect that Phases 2-4 can be performed in any order, even in parallel,
+depending on what is the highest value target.
+
+### **Migration Plan**
+
+Changing the contents and configuration of the shipped AMIs is considered a
+compatibility breaking change, and so needs to be handled with some care. 
+
+We will be rolling out the images over 3 major releases of Teleport:
+
+1. **Teleport 13**: begin publishing hardened images to `teleport-prod` in
+   parallel with legacy images in `gravitational` account.
+2. **Teleport 14**: make hardened images the default, and deprecate (but 
+   still publish) legacy ones
```

Comment:
Just to be clear, we're probably not going to be able to make hardened images the default until the HA Terraform example is overhauled to support a new monitoring stack, along with some of the other stuff described in https://github.com/gravitational/teleport/issues/24785.

Reply:
I think part of the problem here is that we are using these AMIs for at least two different purposes:

1. Delivering Teleport to customers
2. To support our Terraform example

...and currently we can't progress on (1) without also doing a bunch of work on (2) as well.

Should we look at building a baseline "shipping Teleport" AMI, and then layering on the services required for the example in a separate, derived AMI?

Reply:
Maybe. I don't actually know of anywhere in our docs which details how to use the AMIs _without_ our example HA Terraform, though. People are undoubtedly using the AMIs without our Terraform just to save the effort of building their own AMIs, but it's never been something we've actively told people how to do AFAIK.

Back when I overhauled some of these scripts (4 years ago) the idea was to have one AMI to rule them all. It makes sense that this might have changed, so if the idea is that we want to ship a hardened AMI which just has Teleport and nothing else in it, perhaps we could.

This might go along with an idea I discussed with @hugoShaka for "bring your own AMI" - essentially rather than bundle and proscribe a whole AMI for the HA Terraform, we provide a way to install the systemd units and scripts which enable its use into any AMI you like, as long as it has a `teleport` binary available inside it.

The downside of doing it this way is fragmentation and confusion for end users. If we have a "base" hardened AMI for every version of Teleport and then also offer a second "Terraform" AMI on top of this, it makes discoverability a little harder and might lead to people building clusters based on the wrong AMI.

Reply:
After playing with the AWS ImageBuilder for a bit, I can see this as a set of ImageBuilder recipe components... e.g. Here's your "install teleport" component, there is your "enable certbot" components, and so on.